File: python/context_engineering/cache_topology.py
"""
Cache-Topology-Aware Packing

Structures context output to maximize prefix cache hits across API calls.
Anthropic: 90% cost reduction, 85% latency reduction with prefix caching.
OpenAI: 50% cost reduction, automatic for prompts >1024 tokens.

The key insight: if items are always sorted by score, every request produces
a different prefix, destroying cache reuse. Instead, we partition items into
a stable prefix (deterministically ordered) and a volatile suffix (score-ordered).

Based on: https://ankitbko.github.io/blog/2025/08/prompt-engineering-kv-cache/
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Literal, Optional

from .core import Budget, ContextItem, estimate_tokens, pack, pack_async
from .errors import ValidationError

Volatility = Literal["static", "session", "request"]

# Kind-to-volatility mapping
_STATIC_KINDS = {"system", "tool", "schema", "example", "instruction", "few-shot"}
_SESSION_KINDS = {"memory", "conversation", "history", "session"}
_REQUEST_KINDS = {"query", "retrieval", "tool-result", "request"}


@dataclass
class CacheConfig:
    """Provider-specific cache configuration."""

    provider: Optional[str] = None  # "anthropic", "openai", "auto"
    min_prefix_tokens: Optional[int] = None
    mark_breakpoints: bool = False


@dataclass
class CacheAwarePack:
    """Result of cache-aware packing."""

    budget: Budget
    selected: List[ContextItem]
    dropped: List[ContextItem]
    total_tokens: int
    stats: Dict[str, Any]
    cache_key: str
    cacheable_tokens: int
    volatile_tokens: int
    cache_efficiency: float
    partition_boundaries: List[int]


def classify_volatility(item: ContextItem) -> Volatility:
    """Assign a volatility level to an item based on its kind and metadata.

    Items are classified into three tiers:
    - static: system prompts, tool definitions, few-shot examples (rarely change)
    - session: conversation history, memory retrievals (change per session)
    - request: current query, fresh RAG results (change every request)
    """
    # Explicit volatility in metadata takes precedence
    if item.metadata and "volatility" in item.metadata:
        return item.metadata["volatility"]

    kind = (item.kind or "").lower()

    if kind in _STATIC_KINDS:
        return "static"
    if kind in _SESSION_KINDS:
        return "session"
    if kind in _REQUEST_KINDS:
        return "request"

    return "request"


def _hash_string(s: str) -> str:
    """Simple hash for cache key generation."""
    h = 0
    for ch in s:
        h = ((h << 5) - h + ord(ch)) & 0xFFFFFFFF
    return format(h, "x")


def pack_with_cache_topology(
    items: List[ContextItem],
    budget: Budget,
    options: Optional[Dict[str, Any]] = None,
    cache_config: Optional[CacheConfig] = None,
) -> CacheAwarePack:
    """Pack items with cache-topology awareness.

    Partitions items into stable prefix and volatile suffix to maximize
    prefix cache hits across API calls. The stable prefix uses deterministic
    ordering (by id) so the same items always produce the same prefix.

    Args:
        items: Context items to pack (should have `kind` set for best results)
        budget: Token budget
        options: Standard pack options (weights, etc.)
        cache_config: Cache-specific configuration

    Returns:
        CacheAwarePack with cache metadata

    Example:
        result = pack_with_cache_topology(
            [
                ContextItem(id="sys", content="You are helpful.", kind="system", priority=10),
                ContextItem(id="doc", content="Retrieved doc...", kind="retrieval", priority=7),
                ContextItem(id="q", content="User question", kind="query", priority=9),
            ],
            Budget(max_tokens=4096),
        )
        print(result.cache_key)        # stable across requests with same static items
        print(result.cache_efficiency)  # 0.0 - 1.0
    """
    _VALID_PROVIDERS = {"anthropic", "openai", "auto"}
    if cache_config is not None and cache_config.provider is not None:
        if cache_config.provider not in _VALID_PROVIDERS:
            raise ValidationError(
                "Invalid cache config",
                details=[
                    {
                        "path": "cache_config.provider",
                        "message": f"provider must be one of {sorted(_VALID_PROVIDERS)} or None",
                    }
                ],
            )

    config = cache_config or CacheConfig()
    pack_kwargs = dict(options or {})

    # 1. Classify items by volatility
    static_items: List[ContextItem] = []
    session_items: List[ContextItem] = []
    request_items: List[ContextItem] = []

    for item in items:
        v = classify_volatility(item)
        if v == "static":
            static_items.append(item)
        elif v == "session":
            session_items.append(item)
        else:
            request_items.append(item)

    # 2. Sort static items deterministically by id
    static_items.sort(key=lambda x: x.id)

    # 3. Sort session items by recency
    session_items.sort(key=lambda x: x.recency or 0)

    # 4. Pack within each partition
    max_tokens = budget.max_tokens - (budget.reserve_tokens or 0)
    remaining = max_tokens

    # Static items: include all that fit
    selected_static: List[ContextItem] = []
    for item in static_items:
        tokens = item.tokens or estimate_tokens(item.content)
        if tokens <= remaining:
            item_copy = item.model_copy()
            item_copy.tokens = tokens
            selected_static.append(item_copy)
            remaining -= tokens

    # Session items: pack by score within remaining budget
    if remaining > 0 and session_items:
        session_pack = pack(session_items, Budget(max_tokens=remaining), **pack_kwargs)
        session_selected = session_pack.selected
        session_dropped = session_pack.dropped
        remaining -= session_pack.total_tokens
    else:
        session_selected = []
        session_dropped = list(session_items)

    # Request items: pack by score into whatever's left
    if remaining > 0 and request_items:
        request_pack = pack(request_items, Budget(max_tokens=remaining), **pack_kwargs)
        request_selected = request_pack.selected
        request_dropped = request_pack.dropped
    else:
        request_selected = []
        request_dropped = list(request_items)

    # 5. Compose final ordered list: static -> session -> request
    selected = selected_static + session_selected + request_selected
    selected_ids = {i.id for i in selected_static}
    dropped_static = [i for i in static_items if i.id not in selected_ids]
    dropped = dropped_static + session_dropped + request_dropped

    # 6. Add breakpoint markers if configured
    if config.mark_breakpoints:
        static_end = len(selected_static)
        session_end = static_end + len(session_selected)

        if static_end > 0 and static_end < len(selected):
            item = selected[static_end - 1]
            meta = dict(item.metadata or {})
            meta["_cacheBreakpoint"] = "static-end"
            selected[static_end - 1] = item.model_copy(update={"metadata": meta})

        if session_end > static_end and session_end < len(selected):
            item = selected[session_end - 1]
            meta = dict(item.metadata or {})
            meta["_cacheBreakpoint"] = "session-end"
            selected[session_end - 1] = item.model_copy(update={"metadata": meta})

    # 7. Compute cache key from stable prefix content
    static_content = "|".join(f"{i.id}:{i.content}" for i in selected_static)
    cache_key = _hash_string(static_content)

    static_tokens = sum(i.tokens or 0 for i in selected_static)
    total_tokens = sum(i.tokens or 0 for i in selected)

    return CacheAwarePack(
        budget=budget,
        selected=selected,
        dropped=dropped,
        total_tokens=total_tokens,
        stats={
            "staticCount": len(selected_static),
            "sessionCount": len(session_selected),
            "requestCount": len(request_selected),
            "remainingTokens": max(0, max_tokens - total_tokens),
        },
        cache_key=cache_key,
        cacheable_tokens=static_tokens,
        volatile_tokens=total_tokens - static_tokens,
        cache_efficiency=round(static_tokens / total_tokens, 3) if total_tokens > 0 else 0,
        partition_boundaries=[
            len(selected_static),
            len(selected_static) + len(session_selected),
        ],
    )


def _extract_async_kwargs(options: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    """Extract keyword arguments compatible with pack_async from options dict."""
    if not options:
        return {}
    kwargs: Dict[str, Any] = {}
    if "weights" in options:
        kwargs["weights"] = options["weights"]
    if "redundancy_config" in options:
        kwargs["redundancy_config"] = options["redundancy_config"]
    if "provider" in options:
        kwargs["provider"] = options["provider"]
    if "allow_compression" in options:
        kwargs["allow_compression"] = options["allow_compression"]
    return kwargs


async def pack_with_cache_topology_async(
    items: List[ContextItem],
    budget: Budget,
    options: Optional[Dict[str, Any]] = None,
    cache_config: Optional[CacheConfig] = None,
) -> CacheAwarePack:
    """Async version of pack_with_cache_topology that uses pack_async internally.

    Supports asynchronous redundancy elimination via pack_async.
    See pack_with_cache_topology for full documentation.

    Args:
        items: Context items to pack (should have `kind` set for best results)
        budget: Token budget
        options: Standard pack options (weights, etc.)
        cache_config: Cache-specific configuration

    Returns:
        CacheAwarePack with cache metadata
    """
    _VALID_PROVIDERS = {"anthropic", "openai", "auto"}
    if cache_config is not None and cache_config.provider is not None:
        if cache_config.provider not in _VALID_PROVIDERS:
            raise ValidationError(
                "Invalid cache config",
                details=[
                    {
                        "path": "cache_config.provider",
                        "message": f"provider must be one of {sorted(_VALID_PROVIDERS)} or None",
                    }
                ],
            )

    config = cache_config or CacheConfig()
    async_kwargs = _extract_async_kwargs(options)

    # 1. Classify items by volatility
    static_items: List[ContextItem] = []
    session_items: List[ContextItem] = []
    request_items: List[ContextItem] = []

    for item in items:
        v = classify_volatility(item)
        if v == "static":
            static_items.append(item)
        elif v == "session":
            session_items.append(item)
        else:
            request_items.append(item)

    # 2. Sort static items deterministically by id
    static_items.sort(key=lambda x: x.id)

    # 3. Sort session items by recency
    session_items.sort(key=lambda x: x.recency or 0)

    # 4. Pack within each partition
    max_tokens = budget.max_tokens - (budget.reserve_tokens or 0)
    remaining = max_tokens

    # Static items: include all that fit
    selected_static: List[ContextItem] = []
    for item in static_items:
        tokens = item.tokens or estimate_tokens(item.content)
        if tokens <= remaining:
            item_copy = item.model_copy()
            item_copy.tokens = tokens
            selected_static.append(item_copy)
            remaining -= tokens

    # Session items: pack by score within remaining budget
    if remaining > 0 and session_items:
        session_pack = await pack_async(session_items, Budget(max_tokens=remaining), **async_kwargs)
        session_selected = session_pack.selected
        session_dropped = session_pack.dropped
        remaining -= session_pack.total_tokens
    else:
        session_selected = []
        session_dropped = list(session_items)

    # Request items: pack by score into whatever's left
    if remaining > 0 and request_items:
        request_pack = await pack_async(request_items, Budget(max_tokens=remaining), **async_kwargs)
        request_selected = request_pack.selected
        request_dropped = request_pack.dropped
    else:
        request_selected = []
        request_dropped = list(request_items)

    # 5. Compose final ordered list: static -> session -> request
    selected = selected_static + session_selected + request_selected
    selected_ids = {i.id for i in selected_static}
    dropped_static = [i for i in static_items if i.id not in selected_ids]
    dropped = dropped_static + session_dropped + request_dropped

    # 6. Add breakpoint markers if configured
    if config.mark_breakpoints:
        static_end = len(selected_static)
        session_end = static_end + len(session_selected)

        if static_end > 0 and static_end < len(selected):
            item = selected[static_end - 1]
            meta = dict(item.metadata or {})
            meta["_cacheBreakpoint"] = "static-end"
            selected[static_end - 1] = item.model_copy(update={"metadata": meta})

        if session_end > static_end and session_end < len(selected):
            item = selected[session_end - 1]
            meta = dict(item.metadata or {})
            meta["_cacheBreakpoint"] = "session-end"
            selected[session_end - 1] = item.model_copy(update={"metadata": meta})

    # 7. Compute cache key from stable prefix content
    static_content = "|".join(f"{i.id}:{i.content}" for i in selected_static)
    cache_key = _hash_string(static_content)

    static_tokens = sum(i.tokens or 0 for i in selected_static)
    total_tokens = sum(i.tokens or 0 for i in selected)

    return CacheAwarePack(
        budget=budget,
        selected=selected,
        dropped=dropped,
        total_tokens=total_tokens,
        stats={
            "staticCount": len(selected_static),
            "sessionCount": len(session_selected),
            "requestCount": len(request_selected),
            "remainingTokens": max(0, max_tokens - total_tokens),
        },
        cache_key=cache_key,
        cacheable_tokens=static_tokens,
        volatile_tokens=total_tokens - static_tokens,
        cache_efficiency=round(static_tokens / total_tokens, 3) if total_tokens > 0 else 0,
        partition_boundaries=[
            len(selected_static),
            len(selected_static) + len(session_selected),
        ],
    )

File: packages/ce-core/src/cache-topology.ts
/**
 * Cache-Topology-Aware Packing
 *
 * Structures context output to maximize prefix cache hits across API calls.
 * Anthropic: 90% cost reduction, 85% latency reduction with prefix caching.
 * OpenAI: 50% cost reduction, automatic for prompts >1024 tokens.
 *
 * The key insight: if items are always sorted by score, every request produces
 * a different prefix, destroying cache reuse. Instead, we partition items into
 * a stable prefix (deterministically ordered) and a volatile suffix (score-ordered).
 *
 * Based on: https://ankitbko.github.io/blog/2025/08/prompt-engineering-kv-cache/
 */

import type { Budget, ContextItem, ContextPack, PackOptions } from "./types.js";
import { pack, packAsync } from "./pack.js";
import { estimateTokens } from "./estimate.js";
import { hash64 } from "./hash.js";
import { CacheConfigSchema, validateWithSchema } from "./schemas.js";
import type { MaybeAsync } from "./maybe-async.js";
import { chain } from "./maybe-async.js";

/** Volatility level for cache partitioning. */
export type Volatility = "static" | "session" | "request";

/**
 * Provider-specific cache configuration.
 * Different providers have different prefix caching behaviors.
 */
export interface CacheConfig {
  /** Provider name for cache-specific behavior */
  provider?: "anthropic" | "openai" | "auto";
  /** Minimum prefix length for cache eligibility (tokens) */
  minPrefixTokens?: number;
  /** Whether to insert cache breakpoint markers in metadata */
  markBreakpoints?: boolean;
}

/**
 * Result of cache-aware packing, extending ContextPack with cache metadata.
 */
export interface CacheAwarePack extends ContextPack {
  /** Hash of the stable prefix — same hash = cache hit */
  cacheKey: string;
  /** Tokens in the stable prefix (cacheable across requests) */
  cacheableTokens: number;
  /** Tokens in the volatile suffix (changes per request) */
  volatileTokens: number;
  /** Cache efficiency: cacheableTokens / totalTokens */
  cacheEfficiency: number;
  /** Partition boundaries: [staticEnd, sessionEnd] */
  partitionBoundaries: number[];
}

/**
 * Assign a volatility level to an item based on its kind and metadata.
 *
 * Items are classified into three tiers:
 * - static: system prompts, tool definitions, few-shot examples (rarely change)
 * - session: conversation history, memory retrievals (change per session)
 * - request: current query, fresh RAG results (change every request)
 *
 * @param item - The context item to classify
 * @returns The volatility level: "static", "session", or "request"
 *
 * @example
 * ```ts
 * const v = classifyVolatility({ id: "sys", content: "...", kind: "system" });
 * console.log(v); // "static"
 * ```
 */
export function classifyVolatility(item: ContextItem): Volatility {
  // Explicit volatility in metadata takes precedence
  if (item.metadata?.volatility) {
    return item.metadata.volatility as Volatility;
  }

  const kind = item.kind?.toLowerCase() ?? "";

  // Static: system prompts, tools, schemas, examples, instructions
  if (
    kind === "system" ||
    kind === "tool" ||
    kind === "schema" ||
    kind === "example" ||
    kind === "instruction" ||
    kind === "few-shot"
  ) {
    return "static";
  }

  // Request: queries, retrieval results, tool results
  if (
    kind === "query" ||
    kind === "retrieval" ||
    kind === "tool-result" ||
    kind === "request"
  ) {
    return "request";
  }

  // Session: memory, conversation, history
  if (
    kind === "memory" ||
    kind === "conversation" ||
    kind === "history" ||
    kind === "session"
  ) {
    return "session";
  }

  // Default: items without explicit kind are request-level (volatile)
  return "request";
}

/** Function that packs items into a budget, returning sync or async. */
type PackFn = (
  items: ContextItem[],
  budget: Budget,
  options: PackOptions
) => MaybeAsync<ContextPack>;

/**
 * Shared implementation for both sync and async packWithCacheTopology.
 *
 * Uses MaybeAsync + chain() so the sync path never creates Promises,
 * while the async path chains naturally through .then().
 */
function packWithCacheTopologyImpl(
  items: ContextItem[],
  budget: Budget,
  options: PackOptions,
  cacheConfig: CacheConfig,
  packFn: PackFn
): MaybeAsync<CacheAwarePack> {
  validateWithSchema(CacheConfigSchema, cacheConfig, "cacheConfig");

  const estimator = options.tokenEstimator;

  // 1. Classify items by volatility
  const staticItems: ContextItem[] = [];
  const sessionItems: ContextItem[] = [];
  const requestItems: ContextItem[] = [];

  for (const item of items) {
    const volatility = classifyVolatility(item);
    switch (volatility) {
      case "static":
        staticItems.push(item);
        break;
      case "session":
        sessionItems.push(item);
        break;
      case "request":
        requestItems.push(item);
        break;
    }
  }

  // 2. Sort static items deterministically by id (not by score)
  // This ensures the same items always produce the same prefix
  staticItems.sort((a, b) => a.id.localeCompare(b.id));

  // 3. Sort session items by recency (most recent last, for conversation order)
  sessionItems.sort((a, b) => (a.recency ?? 0) - (b.recency ?? 0));

  // 4. Pack within each partition using score-based selection
  const maxTokens = budget.maxTokens - (budget.reserveTokens ?? 0);
  let remaining = maxTokens;

  // Static items: include all that fit (they're high-value and cacheable)
  const selectedStatic: ContextItem[] = [];
  for (const item of staticItems) {
    const tokens = item.tokens ?? estimateTokens(item.content, { estimator });
    if (tokens <= remaining) {
      selectedStatic.push({ ...item, tokens });
      remaining -= tokens;
    }
  }

  // Session items: include by score within remaining budget
  const sessionPackResult: MaybeAsync<ContextPack> =
    remaining > 0
      ? packFn(sessionItems, { maxTokens: remaining }, options)
      : {
          selected: [],
          dropped: sessionItems,
          totalTokens: 0,
          budget,
          stats: {},
        };

  return chain(sessionPackResult, sessionPack => {
    remaining -= sessionPack.totalTokens;

    // Request items: pack by score into whatever's left
    const requestPackResult: MaybeAsync<ContextPack> =
      remaining > 0
        ? packFn(requestItems, { maxTokens: remaining }, options)
        : {
            selected: [],
            dropped: requestItems,
            totalTokens: 0,
            budget,
            stats: {},
          };

    return chain(requestPackResult, requestPack => {
      // 5. Compose final ordered list: static -> session -> request
      const selected = [
        ...selectedStatic,
        ...sessionPack.selected,
        ...requestPack.selected,
      ];

      const selectedStaticIds = new Set(selectedStatic.map(i => i.id));
      const dropped = [
        ...staticItems.filter(i => !selectedStaticIds.has(i.id)),
        ...sessionPack.dropped,
        ...requestPack.dropped,
      ];

      // 6. Add breakpoint markers if configured
      if (cacheConfig.markBreakpoints) {
        const staticEnd = selectedStatic.length;
        const sessionEnd = staticEnd + sessionPack.selected.length;

        if (staticEnd > 0 && staticEnd < selected.length) {
          selected[staticEnd - 1] = {
            ...selected[staticEnd - 1],
            metadata: {
              ...selected[staticEnd - 1].metadata,
              _cacheBreakpoint: "static-end",
            },
          };
        }
        if (sessionEnd > staticEnd && sessionEnd < selected.length) {
          selected[sessionEnd - 1] = {
            ...selected[sessionEnd - 1],
            metadata: {
              ...selected[sessionEnd - 1].metadata,
              _cacheBreakpoint: "session-end",
            },
          };
        }
      }

      // 7. Compute cache key from stable prefix content
      const staticContent = selectedStatic
        .map(i => `${i.id}:${i.content}`)
        .join("|");
      const cacheKey = hash64(staticContent);

      const staticTokens = selectedStatic.reduce(
        (sum, i) => sum + (i.tokens ?? 0),
        0
      );
      const totalTokens = selected.reduce((sum, i) => sum + (i.tokens ?? 0), 0);

      return {
        budget,
        selected,
        dropped,
        totalTokens,
        stats: {
          staticCount: selectedStatic.length,
          sessionCount: sessionPack.selected.length,
          requestCount: requestPack.selected.length,
          remainingTokens: Math.max(0, maxTokens - totalTokens),
        },
        cacheKey,
        cacheableTokens: staticTokens,
        volatileTokens: totalTokens - staticTokens,
        cacheEfficiency:
          totalTokens > 0
            ? Math.round((staticTokens / totalTokens) * 1000) / 1000
            : 0,
        partitionBoundaries: [
          selectedStatic.length,
          selectedStatic.length + sessionPack.selected.length,
        ],
      };
    });
  });
}

/**
 * Pack items with cache-topology awareness.
 *
 * Partitions items into stable prefix and volatile suffix to maximize
 * prefix cache hits across API calls. The stable prefix uses deterministic
 * ordering (by id) so the same items always produce the same prefix.
 *
 * @param items - Context items to pack (should have `kind` set for best results)
 * @param budget - Token budget
 * @param options - Standard pack options
 * @param cacheConfig - Cache-specific configuration
 * @returns CacheAwarePack with cache metadata
 *
 * @example
 * ```ts
 * const result = packWithCacheTopology(
 *   [
 *     { id: "sys", content: "You are helpful.", kind: "system", priority: 10 },
 *     { id: "doc", content: "Retrieved doc...", kind: "retrieval", priority: 7 },
 *     { id: "q", content: "User question", kind: "query", priority: 9 },
 *   ],
 *   { maxTokens: 4096 },
 *   {},
 *   { provider: "anthropic" }
 * );
 * console.log(result.cacheKey);        // stable across requests with same static items
 * console.log(result.cacheEfficiency);  // 0.0 - 1.0
 * ```
 */
export function packWithCacheTopology(
  items: ContextItem[],
  budget: Budget,
  options: PackOptions = {},
  cacheConfig: CacheConfig = {}
): CacheAwarePack {
  return packWithCacheTopologyImpl(
    items,
    budget,
    options,
    cacheConfig,
    pack
  ) as CacheAwarePack;
}

/**
 * Async variant of packWithCacheTopology.
 *
 * Same logic as packWithCacheTopology but delegates to packAsync() internally,
 * supporting async operations like embedding-based redundancy elimination.
 *
 * @param items - Context items to pack (should have `kind` set for best results)
 * @param budget - Token budget
 * @param options - Standard pack options (supports embeddingProvider, redundancyConfig)
 * @param cacheConfig - Cache-specific configuration
 * @returns Promise<CacheAwarePack> with cache metadata
 */
export async function packWithCacheTopologyAsync(
  items: ContextItem[],
  budget: Budget,
  options: PackOptions = {},
  cacheConfig: CacheConfig = {}
): Promise<CacheAwarePack> {
  return packWithCacheTopologyImpl(
    items,
    budget,
    options,
    cacheConfig,
    packAsync
  ) as Promise<CacheAwarePack>;
}
